File: 04_Stack/06_Car_Fleet/0853-car-fleet.cpp
/*
Problem: LeetCode 853 - Car Fleet

Description:
N cars are going to the same destination along a one-lane road. The destination is target miles away.
Each car i has a constant speed speed[i] (in miles per hour), and initial position position[i] miles towards the target along the road.

A car can never pass another car ahead of it, but it can catch up to it and drive bumper to bumper at the same speed.
The distance between these two cars is ignored - they are assumed to have the same position.

A car fleet is some non-empty set of cars driving at the same position and same speed. Note that a single car is also a car fleet.

If a car catches up to a car fleet right at the destination point, it will still be considered as one car fleet.

Intuition:
To determine the number of car fleets that reach the destination, we can simulate the car movement and track the time it takes for each car to reach the destination.

Approach:
1. Create a vector of pairs to store the positions and speeds of the cars.
2. Sort the vector of pairs based on the positions in descending order.
3. Initialize the count of car fleets to 0.
4. Iterate through each car in the sorted vector:
   - Calculate the time it takes for the car to reach the destination (target - position) / speed.
   - If the current car's time is greater than the previous car's time (car fleet is not possible), increment the count of car fleets.
   - Update the previous car's time with the maximum of the current car's time and the previous car's time.
5. Return the count of car fleets.

Time Complexity:
The time complexity is O(n log n), where n is the number of cars. Sorting the vector of pairs takes O(n log n) time.

Space Complexity:
The space complexity is O(n), where n is the number of cars. We store the positions and speeds of the cars in a vector of pairs.
*/

#include<bits/stdc++.h>
using namespace std;

class Solution {
  public:
    int carFleet(int target, vector<int> &position, vector<int> &speed) {
        int n = position.size();
        vector<pair<int, int>> cars;

        // Create vector of pairs to store positions and speeds
        for (int i = 0; i < n; ++i) {
            cars.push_back({position[i], speed[i]});
        }

        // Sort the vector of pairs based on positions in descending order
        sort(cars.begin(), cars.end(), [](const pair<int, int> &a, const pair<int, int> &b) {
            return a.first > b.first;
        });
        int count = 0;
        double prevTime = 0.0;

        for (int i = 0; i < n; ++i) {
            double currTime = static_cast<double>(target - cars[i].first) / cars[i].second;

            if (currTime > prevTime) {
                count++;
                prevTime = currTime;
            }
        }

        return count;
    }
};

File: 04_Stack/06_Car_Fleet/0853-car-fleet.rs
/*
Problem: LeetCode 853 - Car Fleet

Key Idea:
The key idea is to simulate the movement of cars and calculate the time it takes for each car to reach the target.

Approach:
1. Create a vector `cars` of pairs, where each pair represents a car with its position and speed.
2. Sort the `cars` vector by the position of the cars, with the car closest to the target at the end.
3. Initialize a variable `time` to 0, which represents the current time.
4. Initialize a variable `count` to 0, which represents the number of car fleets.
5. Iterate through the `cars` vector from the end to the beginning (from the car closest to the target to the farthest):
   - Calculate the time it takes for the current car to reach the target using the formula: `time = (target - position) as f64 / speed as f64`.
   - If the calculated time is greater than the current `time`, it means a new car fleet is formed. Increment `count` and update `time` to the calculated time.
6. Return the value of `count`, which represents the number of car fleets.

Time Complexity:
O(n log n), where n is the number of cars. Sorting the `cars` vector takes O(n log n) time.

Space Complexity:
O(n), as we store the cars in a vector.
*/

impl Solution {
    pub fn car_fleet(target: i32, position: Vec<i32>, speed: Vec<i32>) -> i32 {
        let mut cars: Vec<(i32, i32)> = position.into_iter().zip(speed.into_iter()).collect();
        cars.sort_by_key(|&(pos, _)| -pos); // Sort by position in descending order.

        let mut time = 0.0;
        let mut count = 0;

        for (pos, spd) in cars {
            let curr_time = (target - pos) as f64 / spd as f64;
            if curr_time > time {
                count += 1;
                time = curr_time;
            }
        }

        count
    }
}
